### PR TITLE
[xpu] INT8 quantization on Intel XPU

### PR DESCRIPTION
**Pull request by zxd1997066**
# Motivation
This feature is one of features listed in the #3576. There is the XMX intrinsic on the Intel GPUs, e.g., Intel Arc B580, to speedup the int8 computation. This PR is target to ensure the int8 quantization works on the XPU and user can use same quantization config with CUDA. 
# Implementation

Since the int8 quantization design of torchAO design is device agnostic,  the only need to do on XPU is to enable the [int_mm on XPU](https://github.com/pytorch/pytorch/pull/157769) and This PR has landed into the PyTorch core since torch-2.9. In this PR, we add the UT test to ensure the entire int quantization workflow on XPU works. 

# Usage
Users can use same quantization config, such as Int8StaticActivationInt8WeightConfig or Int8DynamicActivationInt8WeightConfig, as CUDA to run the fp8 models on XPU. This device dispatched by the PyTorch op.
```
#For dynamic int8 quantization
dynamic_config = Int8DynamicActivationInt8WeightConfig(version=2)
quantize_(model, dynamic_config)

#For static int8 quantization
static_config = Int8StaticActivationInt8WeightConfig(
            scale=int8_input.scale.detach().clone(),
            granularity=granularity,
        )
quantize_(model, static_config)
```

### CODE REVIEW

**Comment by zxd1997066** [comment]
cc @liangan1

**Review by liangan1 — APPROVED** [review]

**Comment by liangan1** [comment]
@xiaowangintel  pls refine the pr desc. You should highlight:

1. The related op has been enabled in the pytorch core, pls add the related pr link. 
2.  For the end user, what is the E2E quantization config on XPU? 

pls refer to https://github.com/pytorch/ao/pull/3332

**Comment by liangan1** [comment]
@zxd1997066  pined memory is N/A to XPU. pls skip these failed cases on XPU related to pined memory.

File: test/dtypes/test_nf4.py [hunk 1]
```diff
@@ -757,7 +757,7 @@ def world_size(self) -> int:
         return 2
 
     @skip_if_lt_x_gpu(2)
-    @unittest.skipIf(not torch.accelerator.is_available(), "Need GPU available")
+    @unittest.skipIf(not torch.cuda.is_available(), "Need GPU available")
```
Comment (liangan1) [inline test/dtypes/test_nf4.py]: nf4 is not related to this PR. pls remove it and submit it in another pr.

Comment (zxd1997066) [reply test/dtypes/test_nf4.py]: done

**Comment by vkuzo** [comment]
what does performance/accuracy look like

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py [hunk 2]
```diff
@@ -60,7 +61,7 @@ def test_creation_and_attributes(self, config):
             self.test_shape[0],
             bias=False,
             dtype=self.dtype,
-            device="cuda",
+            device=_DEVICE,
```
Comment (jerryzh168) [inline test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: what happens when _DEVICE is None?

Comment (jerryzh168) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: I feel a safer way might be to do `_DEVICE = get_current_accelerator_device()` inside the test function since it's guarded by `@unittest.skipIf(not torch.accelerator.is_available(), "Need GPU available")`

also we could think of just removing the None branch for `get_current_accelerator_device` for simplicity and not having to deal with the None case (error out for None)

**Comment by xiaowangintel** [comment]
> what does performance/accuracy look like

@vkuzo  Thanks for your comments. 
For accuracy details as follows:

meta-llama/Llama-3.1-8B-Instruct on Intel GPU and A100 using torchao int8-dynamic-quantization. The model configs as below:
Precision : torch.bfloat16
quantization configuration : Int8DynamicActivationInt8WeightConfig
dataset : wikitext
Result:
The perplexity res:

> Intel GPU: 9.583
> A100: 9.578

We can not provide the performance data directly here due to the internal legal process. But we will provide it once the benchmark tests are ready on XPU and will update the readme in another PR.

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py (hunk 2 above)
Comment (zxd1997066) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: Thanks, Removed None branch for ```get_current_accelerator_device``` https://github.com/pytorch/ao/pull/3500/files#diff-d2a11602a79e83305208472f1abe6a4106f02ce62a7f9524007181813863fcf6L158

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py [hunk 3]
```diff
@@ -38,9 +38,10 @@
         version=2, granularity=PerRow(), act_mapping_type=MappingType.SYMMETRIC
     ),
 ]
+_DEVICE = get_current_accelerator_device()
```
[the remark this answers is not in the document]
Comment (vkuzo) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: it's confusing to have this as a global as cpu is always available and it's no longer clear whether this is cpu or cuda/xpu/etc

**Review by vkuzo — CHANGES_REQUESTED** [review]

File: torchao/utils.py [hunk 4]
```diff
@@ -155,8 +155,6 @@ def get_available_devices():
 def get_current_accelerator_device():
     if torch.accelerator.is_available():
         return torch.accelerator.current_accelerator()
-    else:
```
Comment (jerryzh168) [inline torchao/utils.py]: it will still return None by default, I think there should be an assert here, it might break some tests so probably better to do in a separate PR

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zxd1997066) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: Thanks for your review, moved ```_DEVICE = get_current_accelerator_device()``` inside the test function, because of  ```@unittest.skipIf(not torch.accelerator.is_available(), "Need GPU available")```, it will not be cpu

File: torchao/utils.py (hunk 4 above)
Comment (zxd1997066) [reply torchao/utils.py]: Thanks, will do that in https://github.com/pytorch/ao/pull/3634

**Comment by liangan1** [comment]
@vkuzo according to the semantic of pytorch, cpu is not the accelerator device.

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py (hunk 2 above)
Comment (liangan1) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: @zxd1997066  I suppose we should use _DEVICE = get_current_accelerator_device() inside the test function since it's guarded by @unittest.skipIf(not torch.accelerator.is_available() and remove the NONE of get_current_accelerator_device  in another PR.

Comment (zxd1997066) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: Yes, I already use _DEVICE = get_current_accelerator_device() inside the test function in this PR and submit another PR https://github.com/pytorch/ao/pull/3634 to remove the NONE of get_current_accelerator_device

**Comment by liangan1** [comment]
@jerryzh168  @vkuzo  We have refined it according to comments. pls help to review again. Thanks.

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py [hunk 5]
```diff
@@ -128,7 +129,7 @@ def test_int8_linear_variants(
         )
 
     @common_utils.parametrize("config", INT8_TEST_CONFIGS)
-    @common_utils.parametrize("device", ["cpu", "cuda"])
+    @common_utils.parametrize("device", ["cpu", _DEVICE])
```
[the remark this answers is not in the document]
Comment (vkuzo) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: it's confusing to use persistent state for available devices.  How about something like this?

```python
def get_available_devices() -> List[torch.device]:
    ...

@common_utils.parametrize("device", get_available_devices())
    ...
    # no persistent state to track, device is passed in to the test args, logic is simple
    x = torch.randn(..., device=device)
```

[the remark this answers is not in the document]
Comment (zxd1997066) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: Thanks for your suggestion, updated.

File: torchao/utils.py [hunk 6]
```diff
@@ -153,8 +153,15 @@ def get_available_devices():
 
 
 def get_current_accelerator_device():
+<<<<<<< HEAD
```
Comment (jerryzh168) [inline torchao/utils.py]: this is still here

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py [hunk 7]
```diff
@@ -224,6 +225,7 @@ def test_available_gpu_kernels(self):
             "extern_kernels._int_mm", 1
         ).check_count("triton_poi_fused", 1).run(code[0])
 
+    @unittest.skipIf(not torch.cuda.is_available(), "Need CUDA available")
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: is this because only cuda is supported for the pin_memory test?

File: torchao/utils.py (hunk 6 above)
Comment (zxd1997066) [reply torchao/utils.py]: missed during rebase, just removed it.

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zxd1997066) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: refined the skip for XPU only.

File: test/quantization/quantize_/workflows/int8/test_int8_tensor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zxd1997066) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: Hi @vkuzo , I already updated to @common_utils.parametrize("device", get_available_devices()) https://github.com/pytorch/ao/pull/3500/changes#diff-76d4f5ccf412b0c545cc447dced3b9235139c14230c9fc9c7efa5dfb3fd9c298R139, anything else also need changes?

cc @jerryzh168

[the remark this answers is not in the document]
Comment (jerryzh168) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: please remove `self._DEVICE ` and `_DEVICE` in the file and replace with the same pattern everywhere

[the remark this answers is not in the document]
Comment (zxd1997066) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: Thanks, will update for this.

[the remark this answers is not in the document]
Comment (zxd1997066) [reply test/quantization/quantize_/workflows/int8/test_int8_tensor.py]: seems get_current_accelerator_device() can not be used as ```@common_utils.parametrize("device", [get_current_accelerator_device()])```, as it will raise AssertionError on cpu. Removed ```self._DEVICE```  and ```_DEVICE```, and defined ```device``` inside the test function https://github.com/pytorch/ao/pull/3782
